high_performance_computing/hpc_openmp/02_intro_openmp.md

```diff
@@ -0,0 +1,146 @@
+ 
+Save the following code in `hello_world_omp.c`:
+
+~~~c
+#include <stdio.h>
+#include <omp.h>
+
+int main() {
+    #pragma omp parallel
+    {
+        printf("Hello World!\n");
+    }
+}
+~~~
+
+You'll likely want to compile it using a standard compiler such as `gcc`, although this may depend on your system. To enable the creation of multi-threaded code based on OpenMP directives, pass the `-fopenmp` flag to the compiler. This flag indicates that you're compiling an OpenMP program:
+
+~~~bash
+gcc hello_world_omp.c -o hello_world_omp -fopenmp
+~~~
+
+Before we run the code we also need to indicate how many threads we wish the program to use.
+One way to do this is to specify this using the `OMP_NUM_THREADS` environment variable, e.g.
+
+~~~bash
+export OMP_NUM_THREADS=4
+~~~
+
+Now you can run it just like any other program using the following command:
+
+~~~bash
+./hello_world_omp
+~~~
+
+When you execute the OpenMP program,
+it will display 'Hello World!' multiple times according to the value we entered in `OMP_NUM_THREADS`,
+with each thread in the parallel region executing the `printf` statement concurrently:
+
+~~~
+Hello World!
+Hello World!
+Hello World!
+Hello World!
+~~~
+
+## OpenMP vs. Low-Level Threading APIs (POSIX Threads)
+
+When it comes to parallel programming with threads, there are two main ways to tackle it: the 
+user-friendly OpenMP and the more intricate Low-Level Threading APIs. In this context, **Low-Level 
+Threading APIs**, refer to interfaces like the Portable Operating System Interface (POSIX), which 
+defines a set of standard functions and interfaces for interacting with operating systems. Each 
+approach has its own advantages and considerations, so let's break them down in simpler terms:
+
+OpenMP serves as a gentle introduction to parallel programming. It offers an easy way to parallelize 
+your code without delving into complex technicalities. This makes it ideal for beginners or anyone 
+seeking a straightforward approach. It's like choosing a reliable car that gets you from point A to 
+point B comfortably. 
+
+### Advantages of OpenMP:
+
+- ***User-Friendly:*** OpenMP requires minimal code adjustments, making it a friendly choice for 
+ newcomers.
+- ***Automated Work Distribution:*** It divides tasks among threads, ensuring balanced workloads.
+- ***Wider Applicability:*** OpenMP functions across different systems and programming languages.
+
+For instance, consider a scenario where you have a task that involves doing the same thing over and 
+over again, like processing a bunch of images. With OpenMP, you can split up the work among different 
+threads, so they each handle a part of the task at the same time.
+
+Now, imagine you're a master chef who wants complete control over every ingredient and spice in a 
```

Comment:
I think you're comparing apples and oranges here. Might want to consider the advantages of openmp over other parallel programming libraries (e.g. HPX, Intel TBB, C++ standard library algorithms

I would say that posix threads are just a bad choice for parallel programming full stop, they are more suitable for situations where you *arn't* doing parallel programming, i.e. concurrent / async stuff. Theoretically you could get better performance writing parallel stuff using threads, but I don't think it would be worth the large amount of effort (and additional knowledge you would need to gain) and the sacrifice of portability / maintainability.

Reply:
I would say the advantages of openmp are that it is very easy to get started on, easy to use on existing serial code so is good to migrate an existing serial code to parallel, is a defacto standard for C for parallel programming, but that other options exist on different languages (lots of choice for C++, multiprocessing for python, rayon on rust)

Reply:
Addressed as suggested!